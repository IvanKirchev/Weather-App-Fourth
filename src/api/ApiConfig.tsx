export default {
    openweatherUrl: 'https://api.openweathermap.org',
    // TODO don't store secrets in source control
    openWeatherAppID: '3045bd446337c6e9aa7ef0719614442d'
}