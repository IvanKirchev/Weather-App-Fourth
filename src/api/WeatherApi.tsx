import Forecast from "../components/Forecast";

interface IForecast {
    city: ICity
    list: IHourForecast[]
    days: IDayForecast[]
}

interface IDayForecast {
    list: IHourForecast[]
    id: string
    name: string,
    averageTemp: number
    order: number
    forecastByHours: IHourForecast[]
    icon: string,
    minTemp: number,
    maxTemp: number,
    weatherDescription: string
}

interface IHourForecast {
    dt: number
    main: IMainForecastData
    weather: IWeatherDescription[]
    wind: IWind
    visibility: number
}

interface IMainForecastData {
        temp: number
        feels_like: number
        temp_min: number
        temp_max: number
        pressure: number
        sea_level: number
        grnd_level: number
        humidity: number
        temp_kf: number
}

interface IWeatherDescription {
    id: number
    main: string
    description: string
    icon: string
}

interface IWind {
    speed: number
    deg: number
    gust: number
}

interface ICity {
    country: string
    id: number
    name: string
    population: number
    sunrise: number
    sunset: number
    timezone: number
}

export type { IForecast, IDayForecast, IHourForecast, IMainForecastData, IWeatherDescription, IWind };
