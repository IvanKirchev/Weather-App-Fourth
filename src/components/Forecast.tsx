import react, { Dispatch, SetStateAction, useEffect, useState } from "react"
import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import { styled } from '@mui/material/styles';
import { IDayForecast, IForecast, IHourForecast } from "../api/WeatherApi";
import DayCard from "./DayCard";    
import HourlyForecast from "./HourlyForecast";
import HourlyFullForecast from "./HourlyFullForecast";

interface ForecastProps {
    forecast: IForecast | null
}

function Forecast(props:ForecastProps) {
    const [selectedDay, setSelectedDay] = useState<IDayForecast | null>(null);
    const [selectedHour, setSelectedHour] = useState<IHourForecast | null>(null);
    const [selectedHourIndex, setSelectedHourIndex] = useState<number>(0);

    const bull = (
        <Box
          component="span"
          sx={{ display: 'inline-block', mx: '2px', transform: 'scale(0.8)' }}
        >
          â€¢
        </Box>
    );

    const StyledDayCard = styled(DayCard)(() => ({
        ":hover": "pointer"
    }));

    return (
        <Box sx={{ flexGrow: 1 }}>
            {props.forecast ? <h2>5 Days forcast for {props.forecast?.city.name}</h2> : null }
            <Grid container spacing={2}>
                {props.forecast?.days.map((d) => {
                    return (
                        <StyledDayCard key={d.id} selected={selectedDay?.id == d.id } day={d} setSelectedDay={setSelectedDay} setSelectedHour={setSelectedHour} selectedHourIndex={selectedHourIndex} setSelectedHourIndex={setSelectedHourIndex}/>
                )
                })}
            </Grid>
            { selectedDay ? (
                <>
                    <Grid container spacing={2} marginTop="1%">
                        <Grid item xs>
                            <HourlyFullForecast hourForecast={selectedHour} />
                        </Grid>
                    </Grid>
                    <Grid container spacing={2} marginTop="1%">
                        <Grid item xs>
                            <HourlyForecast selectedHour={selectedHour} setSelectedHour={setSelectedHour}  day={selectedDay} selectedHourIndex={selectedHourIndex} setSelectedHourIndex={setSelectedHourIndex} />
                        </Grid>
                    </Grid>
                </>
            ) : null }
        </Box>
    );
}

export default Forecast;