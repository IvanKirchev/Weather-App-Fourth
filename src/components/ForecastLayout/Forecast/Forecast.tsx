import react, { useState } from "react";
import Grid from "@mui/material/Grid";
import Box from "@mui/material/Box";
import { styled } from "@mui/material/styles";
import { IDayForecast, IForecast, IHourForecast } from "../../../api/WeatherApi";
import DayCard from "./DayCard/DayCard";
import HourSlider from "./HourSlider/HourSlider";
import HourlyForecast from "./HourlyForecast/HourlyForecast";

interface ForecastProps {
  forecast: IForecast | null;
}

function Forecast(props: ForecastProps) {
  const [selectedDay, setSelectedDay] = useState<IDayForecast | null>(null);
  const [selectedHour, setSelectedHour] = useState<IHourForecast | null>(null);
  const [selectedHourIndex, setSelectedHourIndex] = useState<number>(0);

  const ClickableDayCard = styled(DayCard)(() => ({
    ":hover": "pointer",
  }));

  return (
    <Box sx={{ flexGrow: 1 }}>
      {props.forecast ? (
        <h2>5 Days forcast for {props.forecast?.city.name}</h2>
      ) : null}
      <Grid container spacing={2}>
        {props.forecast?.days.map((d) => (
          <ClickableDayCard
            key={d.id}
            selected={selectedDay?.id === d.id}
            day={d}
            setSelectedDay={setSelectedDay}
            setSelectedHour={setSelectedHour}
            selectedHourIndex={selectedHourIndex}
            setSelectedHourIndex={setSelectedHourIndex}
          />
        ))}
      </Grid>
      {selectedDay ? (
        <>
          <Grid container spacing={2} marginTop="1%">
            <Grid item xs>
              <HourlyForecast hourForecast={selectedHour} />
            </Grid>
          </Grid>
          <Grid container spacing={2} marginTop="1%">
            <Grid item xs>
              <HourSlider
                selectedHour={selectedHour}
                setSelectedHour={setSelectedHour}
                day={selectedDay}
                selectedHourIndex={selectedHourIndex}
                setSelectedHourIndex={setSelectedHourIndex}
              />
            </Grid>
          </Grid>
        </>
      ) : null}
    </Box>
  );
}

export default Forecast;
