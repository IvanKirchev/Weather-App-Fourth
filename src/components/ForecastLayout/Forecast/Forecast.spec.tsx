import React from 'react';
import renderer from 'react-test-renderer';
import Forecast from './Forecast';
import { IForecast } from '../../../api/WeatherApi';

const forecastMock: IForecast = {
    city:{
        id:727011,
        name:"Sofia",
        country:"BG",
        population:1152556,
        timezone:7200,
        sunrise:1637212932,
        sunset:1637212932
    },
    days: [
        {
            id:1637236800000,
            name:"Thursday",
            order:0,
            icon:"04d",
            forecastByHours:[],
            weatherDescription:"broken clouds",
            averageTemp:7.445,
            minTemp:0,
            maxTemp:0
        }
    ]
}

it("Should render Forecast component correctly", () => {
    const component = renderer.create(<Forecast forecast={forecastMock}/>)
    expect(component).toMatchSnapshot();
})