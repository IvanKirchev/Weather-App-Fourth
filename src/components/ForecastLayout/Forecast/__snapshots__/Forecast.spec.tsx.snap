// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render Forecast component correctly 1`] = `
<div
  className="MuiBox-root css-i9gxme"
>
  <h2>
    5 Days forcast for 
    Sofia
  </h2>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-mhc70k-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-true css-4jz4pb-MuiGrid-root"
    >
      <div
        className="day-card-box MuiBox-root css-0"
        onClick={[Function]}
      >
        <div
          className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-2v22o2-MuiPaper-root-MuiCard-root"
        >
          <div
            className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
          >
            <p
              className="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-16m3jxf-MuiTypography-root"
            >
              Thursday
            </p>
            <div
              className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              7.4
              Â°C
            </div>
            <p
              className="MuiTypography-root MuiTypography-body1 css-698btc-MuiTypography-root"
            >
              <img
                src="http://openweathermap.org/img/w/04d.png"
              />
            </p>
            <h5
              className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
            >
              broken clouds
            </h5>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
